File: lab_4/g.cpp
#include<bits/stdc++.h>
using namespace std;

struct Node {
    int data;
    Node *left, *right;
};

Node* newNode(int data) {
    Node* node = new Node();
    node->data = data;
    node->left = node->right = NULL;
    return node;
}

Node* insert(Node* node, int data) {
    if(node == NULL) return newNode(data);
    if(data < node->data) node->left = insert(node->left, data);
    else if(data > node->data) node->right = insert(node->right, data);
    return node;
}

int diameter(Node* tree);
int _diameter(Node* tree, int& height);

int main() {
    int n;
    cin >> n;
    Node *root = NULL;
    for(int i = 0; i < n; i++) {
        int x;
        cin >> x;
        if(root == NULL) root = insert(root, x);
        else insert(root, x);
    }
    cout << diameter(root);
    return 0;
}

int diameter(Node* tree) {
    int height = 0;
    return _diameter(tree, height);
}

int _diameter(Node* tree, int& height) {
    int lh = 0, rh = 0;
    int ldiameter = 0, rdiameter = 0;

    if(tree == NULL) {
        height = 0;
        return 0;
    }

    ldiameter = _diameter(tree->left, lh);
    rdiameter = _diameter(tree->right, rh);

    height = max(lh, rh) + 1;

    return max(lh + rh + 1, max(ldiameter, rdiameter));
}

File: lab_4/g.py
class TreeNode:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

class Tree:
    def __init__(self):
        self.root = None
    
    def insert(self, value):
        if self.root is None:
            self.root = TreeNode(value)
        else:
            self._insert_recursive(self.root, value)

    def _insert_recursive(self, current_node, value):
        if value < current_node.value:
            if current_node.left is None:
                current_node.left = TreeNode(value)
            else:
                self._insert_recursive(current_node.left, value)
        else:
            if current_node.right is None:
                current_node.right = TreeNode(value)
            else:
                self._insert_recursive(current_node.right, value)


tree = Tree()
n = input()
valuesToInsert = [int(i) for i in input().split()]

for i in valuesToInsert:
    tree.insert(i)

def findDiameter(root):
    if not root:
        return (0, 0)
    
    left = findDiameter(root.left)
    right = findDiameter(root.right)
    
    currentDiameter = left[0] + right[0]
    maxDiameter = max(left[1], right[1], currentDiameter)
    
    return (max(left[0], right[0]) + 1, maxDiameter)

print(findDiameter(tree.root)[1] + 1)